### PR TITLE
Implemented rotation of logs in SDK

### PR DESCRIPTION
**Pull request by mneuwert**
## Description
Changes related to log-rotation

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/owncloud/ios-app/blob/master/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: ownCloudSDK/Logging/Writers/OCLogFileWriter.m [hunk 1]
```diff
@@ -120,17 +175,85 @@ - (void)appendMessage:(NSString *)message
...
 	{
-		if (![[NSFileManager defaultManager] removeItemAtURL:self.logFileURL error:&error])
+		NSString *fullPath = [directoryPath stringByAppendingPathComponent:filename];
+		NSDictionary* attributes = [self _attributesForPath:fullPath];
+		if(attributes)
+		{
+			NSDate *creationDate = attributes[NSFileCreationDate];
+			int64_t fileSize = [attributes[NSFileSize] longLongValue];
+			OCLogFileRecord *record = [[OCLogFileRecord alloc] initWithName:filename
+															   creationDate:creationDate
```
Comment (felix-schwarz) [inline ownCloudSDK/Logging/Writers/OCLogFileWriter.m]: This entire method and `OCLogFileRecord ` could be greatly simplified:
- using `-[NSFileManager contentsOfDirectoryAtURL:includingPropertiesForkeys:options:error:]` to shallow-enumerate the log folder, you'd get an array of URLs *and* could already pre-fetch attributes like size and creation date in a single line.
- designing `OCLogFileRecord` around and initializing its instances with log file URLs, the class can derive its property values inside the inititializer (or lazily - in property accessors) for size and date through `-[NSURL getResourceValue:forKey:error:]`.

That would remove all the code currently used to put paths together (both here and in `-[OCLogFileRecord fullPath]`) and use the modern URL-based APIs instead.

File: ownCloudSDK/Logging/Writers/OCLogFileWriter.m [hunk 2]
```diff
@@ -144,6 +267,78 @@ - (nullable NSError *)eraseOrTruncate
...
+	{
+		// Check the age of current log
+		NSTimeInterval age = [logCreationDate timeIntervalSinceNow];
+		if (-age > self.rotationInterval)
+		{
+			// Construct path for the archived log
+			NSString *timeStamp = [OCLogWriter timestampStringFrom:[NSDate date]];
+			NSString *transformedTimestamp = [timeStamp stringByReplacingOccurrencesOfString:@"[ /:]+" withString:@"_" options:NSRegularExpressionSearch range:NSMakeRange(0, [timeStamp length])];
+
+			NSString *arhivedLogPath = [self.logFileURL.path stringByAppendingFormat:@".%@", transformedTimestamp];
```
Comment (felix-schwarz) [inline ownCloudSDK/Logging/Writers/OCLogFileWriter.m]: Little typo here: should be ar`c`hivedLogPath.

File: ownCloudSDK/Logging/Writers/OCLogFileWriter.m [hunk 3]
```diff
@@ -144,6 +267,78 @@ - (nullable NSError *)eraseOrTruncate
...
+
+			// Rename current log and start new one
+			[[NSFileManager defaultManager] moveItemAtPath:self.logFileURL.path toPath:arhivedLogPath error:&error];
+
+			// Check if some old logs can be deleted
+			[self cleanUpLogs:NO];
+
+			// Notify observers that contents of the log storage have changed
```
[the remark this answers is not in the document]
Comment (mneuwert) [reply ownCloudSDK/Logging/Writers/OCLogFileWriter.m]: @felix-schwarz you are right but.. still notification is also required to be sent upon log rotation which not always does imply deletion. I refactored the code a bit and kept it where it was and added another invocation in `cleanUpLogs()`

File: ownCloudSDK/Logging/Writers/OCLogFileWriter.m [hunk 4]
```diff
@@ -144,6 +267,78 @@ - (nullable NSError *)eraseOrTruncate
...
+		{
+			[self _scheduleLogRotationTimer];
+		}
+	}
+}
+
 @end
 
+OCIPCNotificationName OCRemoteLogRotationNotificationName = @"org.owncloud.log.rotation";
+NSNotificationName OCLocalLogRotationNotificationName = @"OCLocalLogRotationNotificationName";
```
[the remark this answers is not in the document]
Comment (mneuwert) [reply ownCloudSDK/Logging/Writers/OCLogFileWriter.m]: @felix-schwarz changed

File: ownCloudSDK/Logging/OCLogger.m [hunk 5]
```diff
@@ -518,6 +517,20 @@ - (void)pauseWritersWithIntermittentBlock:(dispatch_block_t)intermittentBlock
...
 }
 
+- (void)pause
+{
+	dispatch_async([OCLogWriter queue], ^{
+		[self _closeAllWriters];
+	});
+}
+
+- (void)resume
```
[the remark this answers is not in the document]
Comment (mneuwert) [reply ownCloudSDK/Logging/OCLogger.m]: @felix-schwarz I though.. a decoupled version of` pause/resume` might be still useful.. but ok, removed

File: ownCloudSDK/App Identity/OCAppIdentity.m [hunk 6]
```diff
@@ -81,6 +82,23 @@ - (NSURL *)appGroupContainerURL
...
+{
+	if (_appGroupLogsContainerURL == nil)
+	{
+		_appGroupLogsContainerURL = [[self appGroupContainerURL] URLByAppendingPathComponent:@"logs"];
+		 if (![[NSFileManager defaultManager] fileExistsAtPath:[_appGroupLogsContainerURL path]])
+		 {
+			 [[NSFileManager defaultManager] createDirectoryAtURL:_appGroupLogsContainerURL
+									  withIntermediateDirectories:NO
+													   attributes:nil
+															error:nil];
```
[the remark this answers is not in the document]
Comment (mneuwert) [reply ownCloudSDK/App Identity/OCAppIdentity.m]: @felix-schwarz I checked and it seems ok

File: ownCloudSDK/Logging/Writers/OCLogFileWriter.m (hunk 2 above)
Comment (mneuwert) [reply ownCloudSDK/Logging/Writers/OCLogFileWriter.m]: @felix-schwarz fixed

File: ownCloudSDK/Logging/OCLogger.m [hunk 7]
```diff
@@ -378,7 +377,7 @@ - (void)rawAppendLogLevel:(OCLogLevel)logLevel functionName:(NSString * _Nullabl
 	}
 	else
 	{
-		dispatch_async(_writerQueue, ^{
+		dispatch_async([OCLogWriter queue], ^{
```
[the remark this answers is not in the document]
Comment (mneuwert) [reply ownCloudSDK/Logging/OCLogger.m]: @felix-schwarz reverted back to previous queue implementation / placement, but exposed it as property in `OCLogger`.

File: ownCloudSDK/Logging/Writers/OCLogFileWriter.m (hunk 1 above)
Comment (mneuwert) [reply ownCloudSDK/Logging/Writers/OCLogFileWriter.m]: @felix-schwarz makes sense.. changed

**Review by felix-schwarz — APPROVED** [review]

File: ownCloudSDK/App Identity/OCAppIdentity.m (hunk 6 above)
[the remark this answers is not in the document]
Comment (felix-schwarz) [reply ownCloudSDK/App Identity/OCAppIdentity.m]: Chances are it's tab widths. That's what it looks like on GitHub for me (which seems to use a tab width of 8 - at least for me):

Bildschirmfoto 2019-05-21 um 09 37 38

Fine with the way it is, though. The rest of the world likely uses a tab width of 2 or 4 and it likely looks ok then.

**Pull request merged**